### PR TITLE
Fix read/write of bits USE_TYPO_METRICS and WWS for OS/2 version < 4

### PR DESCRIPTION
**Pull request by fred-wang**
cc'ing @khaledhosny 

Download Asana Math and open it with fontforge:

http://www.ctan.org/tex-archive/fonts/Asana-Math/

ttx on Asana-Math.ttf generates fsSelection value="00000000 01000000" for the OS/2 table, USE_TYPO_METRICS is *not* set.

Open Asana-Math.ttf with fontforge. In the OS/2 menu, you should see that the option "use typo metrics" *is* checked. Generate a new ttf file (you can uncheck and check the box again if you want). A verification with ttx still indicates that USE_TYPO_METRICS is *not* set.

Analysis of the problem: parsettf.c sets use_typo_metrics to true by default but only overrides with the font value when OS/2 version >= 4. Also tottf.c set the bit USE_TYPO_METRICS in OS/2 table to true only if has version >= 4.

This patch makes use_typo_metrics default to false when loading a ttf file, so that it is not set by default when opening a font with OS/2 version < 4. It also tries to force upgrade to OS/2 when use_typo_metrics has been enabled by the user, so that this bit can be set during file generation. Similarly for info.weight_width_slope_only.

### CODE REVIEW

**Comment by adrientetar** [comment]
Thanks! Are you making a math font?

**Comment by fred-wang** [comment]
No, trying to make existing fonts work with Gecko MathML...

File: fontforge/tottf.c [hunk 1]
```diff
@@ -3333,7 +3333,7 @@ static void setos2(struct os2 *os2,struct alltabs *at, SplineFont *sf,
 	os2->version = 3;
     if ( sf->use_typo_metrics || sf->weight_width_slope_only )
 	os2->version = 4;
-    if ( sf->os2_version!=0 )
+    if ( sf->os2_version > os2->version )
```
Comment (khaledhosny) [inline fontforge/tottf.c]: I’m not sure why this change was needed, as the line above it will correctly set the version to 4 if use_typo_metrics is set.

**Comment by khaledhosny** [comment]
> ttx on Asana-Math.ttf generates `fsSelection value="00000000 01000000"` for the OS/2 table, USE_TYPO_METRICS is not set.

But the 7th bit is set here.

**Comment by fred-wang** [comment]
> But the 7th bit is set here.
https://www.microsoft.com/typography/otspec/os2.htm#fss
It the "7-th" bit counting from index 0 so actually the 8th bit.

File: fontforge/tottf.c (hunk 1 above)
Comment (fred-wang) [inline fontforge/tottf.c]: because the line just below overrides that :-)

Comment (khaledhosny) [inline fontforge/tottf.c]: Right, that is by design the 0 OS/2 version corresponds to “Automatic” in the GUI, where FontForge will chose which fits best the fields being set, but if the user selects a version explicitly it is always used.

Comment (khaledhosny) [inline fontforge/tottf.c]: I think the correct fix here is for the font info dialog to warn the user if he is setting a field not supported by the selected version, not to override it in this low level code. Otherwise it makes no sense to even allow setting OS/2 version explicitly (not necessarily a bad thing) and the support for it can just be dropped.

Comment (fred-wang) [inline fontforge/tottf.c]: Yes, I guess you are right. However, I don't know for the Python interface if we can do such warning.

**Comment by fred-wang** [comment]
@khaledhosny https://www.microsoft.com/typography/otspec/os2.htm#ver actually says that 0 corresponds to TrueType rev 1.5, so I suspect it is actually not quite correct for fontforge to use it for "automatic"...

**Comment by khaledhosny** [comment]
In SMT I simply [dropped the ability for user to select OS/2 version](https://bitbucket.org/sortsmill/sortsmill-tools/commits/e9b229310d96f54b134bf6c7d6cac85f8078f711) and left only the automatic calculation at build time, I suggest that FontForge do the same.

**Comment by be5invis** [comment]
@khaledhosny So if you drop the ability to select version of OS/2 table, maybe you shouold add an option to make it possible to specify `xAvgCharWidth` if you want to produce monospaced fonts used for CMD. I mean that if your font contains CJK glyphs, `xAvgCharWidth` automatically calculated may make the window too wide since CSRSS (and conemu) calculates window width with this value.
